### PR TITLE
Use `LinearOperator`s in `statespace.IBM`

### PR DESCRIPTION
**Pull request by schmidtjonathan**
# In a Nutshell
Use what `probnum.LinearOperator` provides in order to lazify computations in the IBM.

# Detailed Description
This will hopefully speed up some `filtsmooth`-methods significantly, especially wherever large matrix-matrix and matrix-vector products are computed in succession. The use of `LinearOperator`s will save computation by not explicitly computing intermediate results in such scenarios.
For now, this will be implemented for the `IBM` only, since for other `Integrator`s, `LinearOperator` has to be equipped with additional functionality.

# Related Issues
None.

### CODE REVIEW

File: src/probnum/statespace/discrete_transition.py [hunk 1]
```diff
@@ -14,17 +15,8 @@
...
 except ImportError:
-    from functools import lru_cache
-
-    from cached_property import cached_property
-
-try:
-    # functools.cached_property is only available in Python >=3.8
-    from functools import cached_property, lru_cache
-except ImportError:
-    from functools import lru_cache
```
Comment (schmidtjonathan) [inline src/probnum/statespace/discrete_transition.py:20]: I tidied some stuff up while I was at it. This is not related to the rest of the PR.

Comment (pnkraemer) [reply src/probnum/statespace/discrete_transition.py:20]: if these are not used, then great!

File: tests/test_statespace/test_discrete_transition.py [hunk 2]
```diff
@@ -369,3 +368,119 @@ def test_process_noise_cov_cholesky(self):
...
         np.testing.assert_allclose(received, expected)
+
+
+class TestLinearGaussianLinOps:
+    """Test class for linear Gaussians using LinearOperators where possible.
+
+    Also tests that different forward and backward deal correctly when LinearOperators
+    are used.
+    """
+
```
Comment (pnkraemer) [inline tests/test_statespace/test_discrete_transition.py:379]: stupid (perhaps blunt) question: does this test suite is a C&P from the LinearGaussian testcase, just with linops everywhere? Do you think there is a slimmer version of running these tests (in terms of additional parameters in the other test case). If so, that might be a good template of "how to do this kind of linop refactoring in statespace models".

File: tests/test_statespace/test_integrator.py [hunk 3]
```diff
@@ -74,6 +74,54 @@ def integrator(self):
...
+
+    def test_jacobfun(self, some_normal_rv1):
+        expected = self.dg(0.0, some_normal_rv1.mean)
+        received = self.transition.jacobfun(0.0, some_normal_rv1.mean)
+        np.testing.assert_allclose(received.todense(), expected.todense())
+
+    def test_driftmatfun(self):
+        expected = self.G(0.0)
+        received = self.transition.driftmatfun(0.0)
+        np.testing.assert_allclose(received.todense(), expected.todense())
```
Comment (pnkraemer) [inline tests/test_statespace/test_integrator.py:122]: related to the comment above: I think these tests would be much slimmer if transitions would wrap every matrix into a linop, potentially into a dense one.

**Review by pnkraemer — COMMENTED** [review]
Thanks for taking care of this! I left some questions in the diff. Please feel free to ping me offline if this makes it easier to resolve discussions.

File: benchmarks/randprocs.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import scipy.stats
+
+from probnum import config, linops, randprocs, randvars, statespace
+
+
+class MarkovProcessSampling:
+    """Benchmark sampling from Markov processes."""
+
+    param_names = ["lazy_linalg", "num_samples", "len_trajectory", "order", "dimension"]
+    params = [[True, False], [10], [10], [5], [50, 100]]
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply benchmarks/randprocs.py:12]: Ah yes, that was the thing I forgot, because I didn't write it down, as we discussed 🙈 Thanks for reminding me, I'll take that out

File: src/probnum/randprocs/_markov_process.py [hunk 5]
```diff
@@ -70,8 +71,37 @@ def _sample_at_input(
...
+            [
+                self.transition.jointly_transform_base_measure_realization_list_forward(
+                    base_measure_realizations=base_real,
+                    t=args,
+                    initrv=self.initrv,
+                    _diffusion_list=np.ones_like(args[:-1]),
+                )
+                for base_real in base_measure_realizations
+            ]
+        )
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply src/probnum/randprocs/_markov_process.py:104]: This is only insignificantly adapted from (how do you like this paraphrasing of "copy & paste"?) from `generate_samples()` in `generate_samples.py`

[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply src/probnum/randprocs/_markov_process.py:104]: And `generate_samples` now takes a `MarkovProcess` and calls this `sample` function

File: src/probnum/randvars/_normal.py [hunk 6]
```diff
@@ -536,6 +543,11 @@ def _dense_entropy(self) -> np.float_:
             dtype=np.float_,
         )
 
+    def _assure_linop_cov_cholesky(
+        self, damping_factor: FloatArgType
+    ) -> linops.LinearOperator:
+        return linops.aslinop(self.dense_cov_cholesky(damping_factor=damping_factor))
+
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply src/probnum/randvars/_normal.py:550]: I think both names are equally good :)

File: src/probnum/statespace/discrete_transition.py [hunk 7]
```diff
@@ -334,14 +351,22 @@ def _forward_rv_classic(
         new_cov = H @ crosscov + _diffusion * R
         info = {"crosscov": crosscov}
         if compute_gain:
-            gain = scipy.linalg.solve(new_cov.T, crosscov.T, assume_a="sym").T
+            if config.lazy_linalg:
+                gain = (new_cov.T.inv() @ crosscov.T).T
+            else:
+                gain = scipy.linalg.solve(new_cov.T, crosscov.T, assume_a="sym").T
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply src/probnum/statespace/discrete_transition.py:357]: I am not sure about the overhead the creation of the `linops.Matrix` object and the subsequent "densification" would create. I think this is best handled in a separate PR in which we work on creating a unified interface for the `LinearOperator`s. (And this should - as discussed - happen rather sooner than later).

File: tests/test_randvars/test_constant.py [hunk 8]
```diff
@@ -26,6 +26,17 @@ def setUp(self):
     def test_logpdf(self):
         pass
 
+    def test_cov_linops(self):
+        # Ignore scalar support, because in this case, LinOps make no sense.
+        for supp in self.supports[1:]:
+            with self.subTest():
+                with config(lazy_linalg=True):
+                    rv_lo = randvars.Constant(support=supp)
+                    assert isinstance(rv_lo.cov, linops.LinearOperator)
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randvars/test_constant.py:35]: Yes, tbh. Because `Constant.cov_cholesky` directly returns `Constant.cov`

File: tests/test_randprocs/test_markov_process.py [hunk 9]
```diff
@@ -12,3 +16,85 @@ def test_gauss_markov_process_initrv_is_gaussian(
...
+            np.ones(dynamics_dense.dimension),
+            measvar * np.eye(dynamics_dense.dimension),
+        )
+        prior_process_dense = randprocs.MarkovProcess(
+            initarg=time_domain[0], initrv=initrv_dense, transition=dynamics_dense
+        )
+
+        start_dense = time.time()
+        prior_process_dense.sample(rng=rng, args=time_grid)
+        stop_dense = time.time()
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randprocs/test_markov_process.py:47]: Not sure. Might be dangerous due to fluctuations in the amount of computational load of the machine the test is running on. But I'd still like to test the sampling. Any ideas?

File: tests/test_randprocs/test_markov_process.py [hunk 10]
```diff
@@ -12,3 +16,85 @@ def test_gauss_markov_process_initrv_is_gaussian(
...
+        start_linop = time.time()
+        prior_process_linop.sample(rng=rng, args=time_grid)
+        stop_linop = time.time()
+
+    time_linop = stop_linop - start_linop
+
+    print(f"Dense: {time_dense}")
+    print(f"LinOp: {time_linop}")
+
+    assert time_linop < time_dense
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randprocs/test_markov_process.py:76]: See above

File: tests/test_randprocs/test_markov_process.py [hunk 11]
```diff
@@ -12,3 +16,85 @@ def test_gauss_markov_process_initrv_is_gaussian(
...
+
+
+def test_bad_args_shape():
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randprocs/test_markov_process.py:79]: This is the one added line I told you about where codecov proved itself a nuisance, some times. In order to reach the required coverage, I had to check whether a `ValueError` is raised when passing a >1-dimensional time grid to `sample`

File: tests/test_statespace/test_discrete_transition.py (hunk 2 above)
Comment (schmidtjonathan) [reply tests/test_statespace/test_discrete_transition.py:379]: There might be a slimmer version, but it is not straightforward, imo. And in the end some tests might be ran that make no sense. So I figured, this is the easier solution.

File: tests/test_statespace/test_integrator.py (hunk 3 above)
Comment (schmidtjonathan) [reply tests/test_statespace/test_integrator.py:122]: I would like to handle this in a separate PR, together with some extensions and improvements to `linops`.

File: tests/test_randvars/test_normal.py [hunk 12]
```diff
@@ -585,6 +585,25 @@ def test_precompute_cov_cholesky(self):
         with self.subTest("Cholesky is precomputed"):
             self.assertTrue(rv.cov_cholesky_is_precomputed)
 
+    def test_precompute_cov_cholesky_with_linops(self):
```
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randvars/test_normal.py:588]: The method above this one

File: src/probnum/statespace/discrete_transition.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply src/probnum/statespace/discrete_transition.py:357]: I am not sure whether there will be too much overhead, and I don't know what you mean by densification (the input would be dense already? I am talking about having the dense matrix as input to linops.Matrix and then this would be it I guess).

But I think it is fair to do attempt this in a different PR. Just wanted to bring up that I think this is a direction worth going :)

File: tests/test_statespace/test_discrete_transition.py (hunk 2 above)
Comment (pnkraemer) [reply tests/test_statespace/test_discrete_transition.py:379]: Hmm okay thanks for the assessment :) I think it is okay, but was just wondering.

File: tests/test_statespace/test_integrator.py (hunk 3 above)
Comment (pnkraemer) [reply tests/test_statespace/test_integrator.py:122]: okay, sounds good. Let's discuss details offline :)

File: benchmarks/randprocs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply benchmarks/randprocs.py]: Okay cool, please resolve when updated :)

File: src/probnum/randvars/_normal.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply src/probnum/randvars/_normal.py:550]: Hmm okay. My first idea about _assure_... was that this function just checks whether the output is a linop. I didnt expect that it calls dense_cov_cholesky(), which is why I suggested the name change :)

File: tests/test_randprocs/test_markov_process.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tests/test_randprocs/test_markov_process.py:47]: isnt this what the benchmarks are for?

File: tests/test_randprocs/test_markov_process.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tests/test_randprocs/test_markov_process.py:79]: oh god :D Well now we can be sure about this small corner case ;)

File: tests/test_randvars/test_normal.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (pnkraemer) [reply tests/test_randvars/test_normal.py:588]: okay, I was just wondering because there was no diff (I would've expected that the other test runs with `config(lazy_linalg=False)`)

File: src/probnum/randvars/_normal.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply src/probnum/randvars/_normal.py:550]: Ok I can change it

File: tests/test_randvars/test_normal.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randvars/test_normal.py:588]: I made it explicit

File: tests/test_randprocs/test_markov_process.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (schmidtjonathan) [reply tests/test_randprocs/test_markov_process.py:79]: Yep 😆

File: src/probnum/statespace/generate_samples.py [hunk 13]
```diff
@@ -22,12 +18,10 @@ def generate_samples(
     ----------
     rng
         Random number generator.
-    dynmod
-        Transition model describing the prior dynamics.
+    process
+        Markov process to sample from, defining dynamics and initial conditions.
```
Comment (pnkraemer) [inline src/probnum/statespace/generate_samples.py:22]: `process` -> `prior_process`

**Review by pnkraemer — APPROVED** [review]
Thanks a lot for doing this @schmidtjonathan! There is one typo left, but you have my approval anyway.

File: src/probnum/statespace/generate_samples.py (hunk 13 above)
Comment (schmidtjonathan) [reply src/probnum/statespace/generate_samples.py]: Whoopsie! Fixed ✅
